packages/react-router/src/useNavigate.tsx

```diff
@@ -5,36 +5,41 @@ import { LinkOptions, NavigateOptions } from './link'
 import { AnyRoute } from './route'
 import { RoutePaths } from './routeInfo'
 import { RegisteredRouter } from './router'
-import { StringLiteral } from './utils'
 
-export function useNavigate<
+export type UseNavigateResult<TDefaultFrom extends string> = <
   TRouteTree extends AnyRoute = RegisteredRouter['routeTree'],
-  TDefaultFrom extends RoutePaths<TRouteTree> | string = RoutePaths<TRouteTree>,
->(_defaultOpts?: { from?: StringLiteral<TDefaultFrom> }) {
+  TFrom extends RoutePaths<TRouteTree> | string = TDefaultFrom,
+  TTo extends string = '',
+  TMaskFrom extends RoutePaths<TRouteTree> | string = TFrom,
+  TMaskTo extends string = '',
+>({
+  from,
+  ...rest
+}: NavigateOptions<TRouteTree, TFrom, TTo, TMaskFrom, TMaskTo>) => Promise<void>
+
+export function useNavigate<
+  TDefaultFrom extends string = string,
+>(_defaultOpts?: {
+  from?:
+    | TDefaultFrom
```

Comment:
This still provides auto complete without defaulting TFrom to a huge union

Reply:
what exactly is the relevant change here?
would `TDefaultFrom extends RoutePaths<TRouteTree> | string = string,` result in the same improvement?

just trying to understand this better

Reply:
I tried this. I think it works for performance but it breaks autocomplete. I noticed you get autocomplete with:

`TDefaultFrom extends RoutePaths<TRouteTree> | string = RoutePaths<TRouteTree>`. Which might be why it was defaulting to the union in the first place.

`TDefaultFrom | RoutePaths<RouteTree> | (string & {})` provides loose autocomplete. But hopefully this gets fixed with just `TDefaultFrom | RoutePaths<RouteTree> | string` soon 

https://www.totaltypescript.com/typescript-5-3#automatic-loose-autocomplete-on-strings

This method keeps the TDefaultFrom looser but still provides autocomplete which I think is what you want here because the constraint was `RoutePaths<TRouteTree> | string` which is already very loose due to the `| string`

Anyway. The point is to remove the default to a large union but to keep autocomplete. I think defaulting this to a large union could easily sneak in performance issues as you just need to do one `extends` and you're dealing with it being distributed somewhere else and it could be hard to track this down.

Reply:
I will also explain a separate issue which is connected to this but it was why `useNavigate` seemed so slow. `useNavigate` captured `TRouteTree` as a type parameter and this was used in `navigate` which is a function you return. It looks like Typescript was checking `TRouteTree` again when `navigate` was called. If I let `navigate` capture `TRouteTree` instead, the extra checking didn't happen. With a large tree this checking was taking seconds for me. I'm not sure why but this might be to do with generic constraints on `TRouteTree` and variance. But `TRouteTree` will never change so this checking was only slowing things down. Especially if the `TRouteTree` is large